cmd/gateway-azure.go

```diff
@@ -57,29 +57,21 @@ const metadataObjectNameTemplate = globalMinioSysTmp + "multipart/v1/%s.%x/azure
 func s3MetaToAzureProperties(s3Metadata map[string]string) (storage.BlobMetadata,
 	storage.BlobProperties) {
 
-	// Azure does not permit user-defined metadata key names to
-	// contain hyphens. So we map hyphens to underscores for
-	// encryption headers. More such headers may be added in the
-	// future.
-	gatewayHeaders := map[string]string{
-		"X-Amz-Meta-X-Amz-Key":     "X-Amz-Meta-x_minio_key",
-		"X-Amz-Meta-X-Amz-Matdesc": "X-Amz-Meta-x_minio_matdesc",
-		"X-Amz-Meta-X-Amz-Iv":      "X-Amz-Meta-x_minio_iv",
-	}
-
 	var blobMeta storage.BlobMetadata = make(map[string]string)
 	var props storage.BlobProperties
 	for k, v := range s3Metadata {
 		k = http.CanonicalHeaderKey(k)
-		if nk, ok := gatewayHeaders[k]; ok {
-			k = nk
-		}
 		switch {
 		case strings.HasPrefix(k, "X-Amz-Meta-"):
 			// Strip header prefix, to let Azure SDK
 			// handle it for storage.
 			k = strings.Replace(k, "X-Amz-Meta-", "", 1)
-			blobMeta[k] = v
+			if strings.Contains(k, "_") {
+				k = strings.Replace(k, "_", "__", -1)
+			} else {
+				k = strings.Replace(k, "-", "_", -1)
```

Comment:
You would need to do `_` and `-` translation irrespective of the condition i.e. it should work for `X-Amz-Meta-My-Metadata_Name`